File: Problems/Week_3/4.7_랜선_자르기(1654)/임용식/1654.cs
using System;

namespace AlgoCSharp
{
    class MainProgram
    {
        static void Main(string[] args)
        {
            int[] input     = Array.ConvertAll(Console.ReadLine().Split(), Convert.ToInt32);
            long[] lan      = new long[input[0]];
            long minLength  = 1;
            long maxLength  = 0;

            for (int i = 0; i < lan.Length; i++)
            {
                lan[i] = Convert.ToInt64(Console.ReadLine().Split()[0]);
                maxLength = maxLength < lan[i] ? lan[i] : maxLength;
            }

            Console.WriteLine(GetMaxCutLength(lan, minLength, maxLength, (long)input[1]));
        }

        static long GetMaxCutLength(long[] lan, long minLength, long maxLength, long lanCount)
        {
            long cuttedLanCount = 0;
            if (minLength > maxLength)
            {
                return maxLength;
            }
            long cutPoint = (minLength + maxLength)/2;

            for (int i = 0; i < lan.Length; i++)
            {
                cuttedLanCount += lan[i] / cutPoint;
            }

            if (cuttedLanCount >= lanCount)
            {
                minLength = cutPoint+1;
            }
            else
            {
                maxLength = cutPoint-1;
            }
            return GetMaxCutLength(lan, minLength, maxLength, lanCount);
        }
    }
}

File: Problems/Week_3/4.7_랜선_자르기(1654)/임용식/1654.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main()
{
	int k, n;
	std::cin >> k >> n;

	std::vector<int64_t> lan(k);

	for (auto& i : lan)
	{
		std::cin >> i;
	}

	int64_t minLength = 1, maxLength = *max_element(lan.begin(), lan.end());

	while (minLength <= maxLength)
	{
		int64_t cutLanCount = 0;
		int64_t cutPoint = (minLength + maxLength) / 2;

		for (const auto& i : lan)
		{
			cutLanCount += i/cutPoint;
		}

		if (cutLanCount >= n)
		{
			minLength = cutPoint + 1;
		}
		else
		{
			maxLength = cutPoint - 1;
		}
	}
	std::cout << maxLength;

	return 0;
}